redis/connection.py

```diff
@@ -734,6 +715,188 @@ def _host_error(self):
         return f"{self.host}:{self.port}"
 
 
+def ensure_string(key):
+    if isinstance(key, bytes):
+        return key.decode("utf-8")
+    elif isinstance(key, str):
+        return key
+    else:
+        raise TypeError("Key must be either a string or bytes")
+
+
+class CacheProxyConnection(ConnectionInterface):
+    def __init__(
+        self,
+        conn: ConnectionInterface,
+        cache: Cache,
+        conf: CacheConfiguration,
+        cache_lock: threading.Lock,
+    ):
+        self.pid = os.getpid()
+        self._conn = conn
+        self.retry = self._conn.retry
+        self.host = self._conn.host
+        self.port = self._conn.port
+        self._cache = cache
+        self._cache_lock = cache_lock
```

Comment:
Guarding cache object calls with an "external" lock makes it impossible to replace the current Cache implementation with something more efficient.

Reply:
Moved lock into the connection itself